From ad2797214c14591193dcd170a26c829257a60aee Mon Sep 17 00:00:00 2001
From: Ulf Hermann <ulf.hermann@qt.io>
Date: Fri, 10 Oct 2025 09:14:34 +0200
Subject: [PATCH 149/239] QtQml: Use QQmlStringConverters to create QColor from
 strings

We use the same method in QQmlPropertyValidators. It's more efficient
and it works even if you haven't imported QtQuick. Using a different
method for the property assignment risks producing an invalid result.

Amends commit a09ea6b3d6265a98c17e190db007fecfbd76f317.

Pick-to: 6.8
Change-Id: I7dda609d0b6a3f9a32c71649a524bf1101d9fd12
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
Reviewed-by: Mitch Curtis <mitch.curtis@qt.io>
(cherry picked from commit fb3d4480962d7929038c9add3405b35f36a936a9)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qml/qml/qqmlobjectcreator.cpp                | 10 +++++-----
 .../qml/qqmllanguage/data/colorWithoutQuick.qml  |  7 +++++++
 tests/auto/qml/qqmllanguage/tst_qqmllanguage.cpp | 16 ++++++++++++++++
 3 files changed, 28 insertions(+), 5 deletions(-)
 create mode 100644 tests/auto/qml/qqmllanguage/data/colorWithoutQuick.qml

diff --git a/src/qml/qml/qqmlobjectcreator.cpp b/src/qml/qml/qqmlobjectcreator.cpp
index 38c9b13c01..5142fd5e03 100644
--- a/src/qml/qml/qqmlobjectcreator.cpp
+++ b/src/qml/qml/qqmlobjectcreator.cpp
@@ -496,11 +496,11 @@ void QQmlObjectCreator::setPropertyValue(const QQmlPropertyData *property, const
     }
     break;
     case QMetaType::QColor: {
-        QVariant data = QQmlValueTypeProvider::createValueType(
-                    compilationUnit->bindingValueAsString(binding), propertyType);
-        if (data.isValid()) {
-            property->writeProperty(_qobject, data.data(), propertyWriteFlags);
-        }
+        bool ok = false;
+        QVariant data = QQmlStringConverters::colorFromString(
+                compilationUnit->bindingValueAsString(binding), &ok);
+        Q_ASSERT(ok); // We've checked this in QQmlPropertyValidator
+        property->writeProperty(_qobject, data.data(), propertyWriteFlags);
     }
     break;
 #if QT_CONFIG(datestring)
diff --git a/tests/auto/qml/qqmllanguage/data/colorWithoutQuick.qml b/tests/auto/qml/qqmllanguage/data/colorWithoutQuick.qml
new file mode 100644
index 0000000000..0b4a35615f
--- /dev/null
+++ b/tests/auto/qml/qqmllanguage/data/colorWithoutQuick.qml
@@ -0,0 +1,7 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR GPL-3.0-only
+import Test
+
+MyTypeObject {
+    colorProperty: "grey"
+}
diff --git a/tests/auto/qml/qqmllanguage/tst_qqmllanguage.cpp b/tests/auto/qml/qqmllanguage/tst_qqmllanguage.cpp
index 6e13828b47..4823696505 100644
--- a/tests/auto/qml/qqmllanguage/tst_qqmllanguage.cpp
+++ b/tests/auto/qml/qqmllanguage/tst_qqmllanguage.cpp
@@ -533,6 +533,8 @@ private slots:
 
     void variantAssociationHasOwnProperty();
 
+    void colorWithoutQuick();
+
 private:
     QQmlEngine engine;
     QStringList defaultImportPathList;
@@ -10068,6 +10070,20 @@ void tst_qqmllanguage::variantAssociationHasOwnProperty()
     QVERIFY(!engine.evaluate("qobject.variantobj.hasOwnProperty('key3')").toBool());
 }
 
+void tst_qqmllanguage::colorWithoutQuick()
+{
+    QQmlEngine engine;
+    QQmlComponent c(&engine, testFileUrl("colorWithoutQuick.qml"));
+    QVERIFY2(c.isReady(), qPrintable(c.errorString()));
+    QScopedPointer<QObject> o(c.create());
+    QVERIFY(!o.isNull());
+    MyTypeObject *t = qobject_cast<MyTypeObject *>(o.data());
+    QVERIFY(t);
+    const QColor expected = QColor::fromString("grey");
+    QVERIFY(expected.isValid());
+    QCOMPARE(t->colorProperty(), expected);
+}
+
 QTEST_MAIN(tst_qqmllanguage)
 
 #include "tst_qqmllanguage.moc"
-- 
2.51.2

